File: flower/src/main/java/com/kharitonov/flower/services/UserService.java
package com.kharitonov.flower.services;

import com.kharitonov.flower.dtos.UserDTO;
import com.kharitonov.flower.dtos.factories.UserDTOFactory;
import com.kharitonov.flower.entities.UserEntity;
import com.kharitonov.flower.exceptions.BadRequestException;
import com.kharitonov.flower.exceptions.NotFoundException;
import com.kharitonov.flower.repositories.UserRepository;
import lombok.NonNull;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class UserService {

    private final UserRepository userRepository;

    @Autowired
    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public Optional<UserEntity> getByLogin(@NonNull String login) {
        return userRepository.findByLogin(login);
    }

    public UserDTO findById(Long id) {
        return new UserDTOFactory().toUserDTO(userRepository.findById(id).orElseThrow(NotFoundException::new));
    }

    public UserDTO save(Optional<UserEntity> userEntity) {
        if(userEntity.isEmpty()) {
            throw new BadRequestException("Object was not found.");
        }

        return new UserDTOFactory().toUserDTO(userRepository.save(userEntity.get()));
    }

    public UserDTO update(Optional<UserEntity> userEntityFromDB, Optional<UserEntity> userEntity) {
        if(userEntityFromDB.isEmpty() || userEntity.isEmpty()) {
            throw new BadRequestException("Object was not found.");
        }
        BeanUtils.copyProperties(userEntity.get(), userEntityFromDB.get(), "id");

        return new UserDTOFactory().toUserDTO(userRepository.save(userEntityFromDB.get()));
    }

    public void delete(Optional<UserEntity> userEntity) {
        if(userEntity.isEmpty()) {
            throw new BadRequestException(("Object was not found."));
        }
        userRepository.delete(userEntity.get());
    }
}

File: src/main/java/by/hghgrtut/shop/services/UserService.kt
package by.hghgrtut.shop.services

import by.hghgrtut.shop.model.UserModel
import by.hghgrtut.shop.model.factories.UserModelFactory
import by.hghgrtut.shop.entities.UserEntity
import by.hghgrtut.shop.exceptions.BadRequestException
import by.hghgrtut.shop.exceptions.NotFoundException
import by.hghgrtut.shop.repositories.UserRepository
import org.springframework.beans.BeanUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.Optional

@Service
class UserService @Autowired constructor(private val userRepository: UserRepository) {
    
    fun getByLogin(login: String): UserEntity? = userRepository.findByLogin(login)

    fun findById(id: Long?): UserModel =
            UserModelFactory().toUserModel(userRepository.findById(id).orElseThrow { NotFoundException() })

    fun save(userEntity: UserEntity?): UserModel = UserModelFactory().toUserModel(userRepository.save(userEntity!!))

    fun update(userEntityFromDB: UserEntity?, userEntity: UserEntity?): UserModel {
        BeanUtils.copyProperties(userEntity!!, userEntityFromDB!!, "id")
        return UserModelFactory().toUserModel(userRepository.save(userEntityFromDB!!))
    }

    fun delete(userEntity: UserEntity?) = userRepository.delete(userEntity!!)
}